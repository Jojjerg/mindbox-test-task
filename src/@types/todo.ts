export interface ITodo {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

export type TodoContextType = {
  todos: ITodo[];
  filter: IFilter;
  getTodos: () => void;
  addTodo: (title: string) => void;
  updateTodo: (id: number) => void;
  deleteCompletedTodos: () => void;
  changeFilter: (filter: IFilter) => void;
};

export type IFilter = "All" | "Completed" | "Active";
