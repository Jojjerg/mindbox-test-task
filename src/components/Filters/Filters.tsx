import type { IFilter } from "../../@types/todo";
import React from "react";

const Filters = ({ changeFilter }: { changeFilter: (filter: IFilter) => void }) => {
  return (
    <div className="flex items-center gap-2">
      <button
        data-testid="filter-button"
        onClick={() => changeFilter("All")}
        className="px-2 border border-transparent border-r-zinc-600"
      >
        <span>Все</span>
      </button>
      <button
        data-testid="filter-button"
        onClick={() => changeFilter("Active")}
        className="px-2"
      >
        <span>Активные</span>
      </button>
      <button
        data-testid="filter-button"
        onClick={() => changeFilter("Completed")}
        className="px-2 border border-transparent border-l-zinc-600"
      >
        <span>Завершенные</span>
      </button>
    </div>
  );
};

export default Filters;
