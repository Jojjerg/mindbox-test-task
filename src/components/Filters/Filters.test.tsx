import { ITodo, TodoContextType } from "../../@types/todo";
import {
  fireEvent,
  getAllByTestId,
  render,
  screen,
} from "@testing-library/react";

import Filters from "./Filters";
import React from "react";
import { act } from "react-test-renderer";
import { createRenderer } from "react-test-renderer/shallow";

const renderer = createRenderer();

let filter = "All";

const changeFilter = (newFilter: TodoContextType["filter"]) => {
  return (filter = newFilter);
};

const defaultComponent = <Filters changeFilter={changeFilter} />;

describe("<Filters />", () => {
  it("should render and match the snapshot", () => {
    renderer.render(defaultComponent);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it("filter should change", async () => {
    const { getAllByTestId } = render(defaultComponent);
    const filterButtons = getAllByTestId("filter-button");

    expect(filterButtons.length).toEqual(3);

    await act(async () => {
      fireEvent.click(filterButtons[0]);
    });

    expect(filter).toBe("All");

    await act(async () => {
      fireEvent.click(filterButtons[1]);
    });

    expect(filter).toBe("Active");

    await act(async () => {
      fireEvent.click(filterButtons[2]);
    });

    expect(filter).toBe("Completed");
  });
});
