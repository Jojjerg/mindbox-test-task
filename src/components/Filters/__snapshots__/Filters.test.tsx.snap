// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filters /> should render and match the snapshot 1`] = `
<div
  className="flex items-center gap-2"
>
  <button
    className="px-2 border border-transparent border-r-zinc-600"
    data-testid="filter-button"
    onClick={[Function]}
  >
    <span>
      Все
    </span>
  </button>
  <button
    className="px-2"
    data-testid="filter-button"
    onClick={[Function]}
  >
    <span>
      Активные
    </span>
  </button>
  <button
    className="px-2 border border-transparent border-l-zinc-600"
    data-testid="filter-button"
    onClick={[Function]}
  >
    <span>
      Завершенные
    </span>
  </button>
</div>
`;
